packages/wrangler/src/config/validation-pages.ts

```diff
@@ -0,0 +1,126 @@
+import type {
+	Config,
+	RawConfig,
+} from "./config";
+
+const supportedPagesConfigFields = [
+	"pages_build_output_dir",
+	"name",
+	"compatibility_date",
+	"compatibility_flags",
+	"placement",
+	"vars",
+	"kv_namespaces",
+	"queues", // `producers` ONLY
+	"r2_buckets",
+	"d1_databases",
+	"vectorize",
+	"services",
+	"analytics_engine_datasets",
+	"ai",
+	"ip",
+	"port",
+	"inspector_port",
+	"local_protocol",
+	"upstream_protocol",
+	"host"
+] as const;
+
+export function validatePagesConfig(
+	rawConfig: RawConfig,
+	config: Config
+): Diagnostics {
+	const diagnostics = new Diagnostics(
+		`Configuration file belonging to ⚡️ Pages ⚡️ project detected.
+		Running Pages configuration validation:`
+	);
+
+	validateMainField(config, diagnostics);
+	validatePagesEnvironmentNames(rawConfig, diagnostics);
+	validateUnsupportedFields(config, diagnostics);
+
+	return diagnostics;
+}
+
+/**
+ * Validate that configuration file doesn't declare "main", if
+ * "pages_build_output_dir" is present
+ */
+function validateMainField(config: Config, diagnostics: Diagnostics) {
+	if (hasProperty(config, "main")) {
+		diagnostics.errors.push(
+			`Configuration file cannot contain both both "main" and "pages_build_output_dir" configuration keys.
+			Please use "main" if you are deploying a Worker, or "pages_build_output_dir" if you are deploying a Pages project.`
+		);
+	}
+}
+
+/**
+ * Validate that no named-environments other than "preview" and "production"
+ * were declared in the configuration file for Pages
+ */
+function validatePagesEnvironmentNames(
+	rawConfig: RawConfig,
+	diagnostics: Diagnostics
+) {
+	if (!rawConfig.env) return;
+
+	const envNames = Object.keys(rawConfig.env);
+	const unsupportedPagesEnvNames = envNames.filter(
+		(name) => name !== "preview" && name !== "production"
+	);
+
+	if (unsupportedPagesEnvNames.length > 0) {
+		diagnostics.errors.push(
+			`Configuration file contains environment names that are not supported by Pages projects:
+			${JSON.stringify(unsupportedPagesEnvNames)}.
+			The supported named-environments for Pages are "preview" and "production".`
+		);
+	}
+}
+
+/**
+ * Check for configuration fields that are not supported by Pages via the
+ * configuration file
+ */
+function validateUnsupportedFields(
+	config: Config,
+	diagnostics: Diagnostics
+) {
+	const unsupportedFields = new Set(Object.keys(config));
+
+	for (const field of supportedPagesConfigFields) {
+		// because only `queues.producers` are supported, but let's not
+		// validate and potentially error here
+		if(field === "queues") { continue; }
+		
+		unsupportedFields.delete(field);
+	}
+
+	if (unsupportedFields.size > 0) {
+		const fields = Array.from(unsupportedFields.keys()).map((field) => `"${field}"`);
+
+		fields.forEach((field) => {
+			if(field === "queues" && !!config.queues?.consumers?.length) {
+				diagnostics.errors.push(
+					`Pages configuration file does not support "queues.consumers"`
+				);
+			} else {
```

Comment:
Why not just do this in the for loop above to avoid having to check for `field == "queues"` twice?

Reply:
because that would break the logical grouping of the keys in `supportedPagesConfigFields` when errors would be printed to stdout, assuming multiple key-related errs. 

So for example if we had errors for each of `compatibility_flags`, `placement`, `kv_namespaces`, `queues` and `d1_databases`, the `queues` related error would always be printed first, which felt off to me. While not a big deal, and I understand not everybody is as OCD as I am 😅 , I figured the user experience would be much nicer, if we can maintain some logical grouping of the errors, based on the config grouping we have in our docs.

Reply:
Fair enough. Should have guessed you had thought about it

Reply:
totes valid callout!

Reply:
I'll add a comment to clarify the "why"

Reply:
✅